File: python/src/basecamp/oauth/device_authorization.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class DeviceAuthorization:
    """RFC 8628 §3.2 device authorization response.

    Returned by :func:`~basecamp.oauth.device.request_device_authorization`. The
    ``device_code`` is polled at the token endpoint; ``user_code`` and
    ``verification_uri`` (optionally ``verification_uri_complete``) are shown to
    the user. ``interval`` defaults to 5 seconds when the server omits it.
    """

    device_code: str
    user_code: str
    verification_uri: str
    expires_in: int
    interval: int
    verification_uri_complete: str | None = None

File: ruby/lib/basecamp/oauth/device_authorization.rb
# frozen_string_literal: true

module Basecamp
  module Oauth
    # RFC 8628 device authorization response (the device/user code pair the user
    # approves out of band).
    #
    # @attr device_code [String] The device verification code polled at the token
    #   endpoint.
    # @attr user_code [String] The code the user enters at the verification URI.
    # @attr verification_uri [String] Where the user goes to approve the request.
    # @attr verification_uri_complete [String, nil] The verification URI with the
    #   user code pre-filled (optional; absent when the server omits it).
    # @attr expires_in [Integer] Lifetime of the device/user codes in seconds.
    # @attr interval [Integer] Minimum seconds to wait between token polls
    #   (defaults to 5 when the server omits it).
    DeviceAuthorization = Data.define(
      :device_code,
      :user_code,
      :verification_uri,
      :verification_uri_complete,
      :expires_in,
      :interval
    ) do
      def initialize(
        device_code:,
        user_code:,
        verification_uri:,
        expires_in:,
        interval:,
        verification_uri_complete: nil
      )
        super
      end
    end
  end
end
